### DISCUSSION TITLE
Assessment of the difficulty in porting CPU architecture for oneTBB

### DISCUSSION THREAD

**Discussion by wangyuliu** [oneapi-src/oneTBB#1204] [Design discussions]
Hello everyone! I am working on implementing a tool to assess the complexity of CPU architecture porting. It primarily focuses on RISC-V architecture porting. In fact, the tool may have an average estimate of various architecture porting efforts.My focus is on the overall workload and difficulty of transplantation in the past and future,even if a project has already been ported.As part of my dataset, I have collected the **oneTBB** project. **I would like to gather community opinions to support my assessment. I appreciate your help and response!** Based on scanning tools, the porting complexity is determined to be moderate leaning towards simple, with a moderate amount of code related to the CPU architecture in the project.  Is this assessment accurate?Do you have any opinions on personnel allocation and consumption time？ I look forward to your help and response.

**Comment by cosmicexplorer** [comment]
In a similar vein, I would also like to understand the hardware requirements more thoroughly for processors implementing some form of the `x86{,_64}` ISA. I found some issues, described below.

# Cyclic Documentation Graph
`SYSTEM_REQUIREMENTS.md` states the following: https://github.com/uxlfoundation/oneTBB/blob/master/SYSTEM_REQUIREMENTS.md#supported-hardware
> # Supported Hardware
>
> - Intel(R) Celeron(R) processor family
> - Intel(R) Core* processor family
> - Intel(R) Xeon(R) processor family
> - Intel(R) Atom* processor family
> - Non-Intel(R) processors compatible with the processors listed above
>
> Changes to the list of supported hardware will be announced in advance with a deprecation period preceding removal or reduction of support.

That document makes no further mention of *hardware* support at all, only covering software platforms. More generally, it conflates "platforms and target triples that the project may be used for" with "platforms and target triples for which we guarantee support for enterprise customers" (while also mentioning "community-supported" platforms, but providing no link or discussion of what "community support" means). This is of course quite strange, given that [`SUPPORT.md`](https://github.com/uxlfoundation/oneTBB/blob/master/SUPPORT.md) already exists, but has the contents one would expect from a `CONTRIBUTING.md`. And `CONTRIBUTING.md` exists, but mentions a totally different set of "prerequisites", with links going out of the repo: https://github.com/uxlfoundation/oneTBB/blob/master/CONTRIBUTING.md#prerequisites
> # Prerequisites
>
> As a contributor, you'll want to be familiar with the oneTBB project and the repository layout. You should also know how to use it as explained in the [oneTBB documentation](https://uxlfoundation.github.io/oneTBB/) and how to set up your build development environment to configure, build, and test oneTBB as explained in the [oneTBB Build System Description](https://github.com/uxlfoundation/oneTBB/blob/master/cmake/README.md).

These references form a lengthy cyclic path: https://uxlfoundation.github.io/oneTBB/GSG/system_requirements.html
> oneAPI Threading Building Blocks (oneTBB) supports a broad list of hardware platforms, operating systems, and compilers. For details, see [oneTBB System Requirements](https://github.com/uxlfoundation/oneTBB/blob/master/SYSTEM_REQUIREMENTS.md).

...which goes back to the `SYSTEM_REQUIREMENTS.md` we started with, which has absolutely no discussion of hardware support.

# BLAKE3 reverse dependency

- In blake3-team/blake3#445 @silvanshade implemented thread-based parallelism for the BLAKE3 tree hash.
    - This is quite exciting, but becomes much less useful if the hardware requirements are unclear.
- In blake3-team/blake3#457 @ibmibmibm demonstrates OpenMP-based parallelism.
    - There is a concern about performance, and of course oneAPI describes its decoupled parallelism configuration interface as an alternative to OpenMP.
    - But if oneTBB is only compatible with Intel-based processors, then an OpenMP alternative would seem to be useful regardless of a claimed performance difference.

This uncertainty is the problem I'm looking to solve.

## practical issues for packaging

As of now, I am unable to package BLAKE3's recent multithreaded support within the spack package manager as a result of this confusion, and I am inclined to advocate for the OpenMP approach to the BLAKE3 package, in order to improve our hardware support.

# forced into alternatives

Without a clear description of oneTBB's compatibility with x86 processors, I am somewhat forced to consider an alternative implementation using something like e.g. [RAJA](https://github.com/llnl/raja), which is very serious about portability and is not likely to break anyone's build on a non-Intel x86 CPU.

Clarity on the portability goals for oneTBB would improve this situation. It would also lay out a path for further architectures such as RISC-V, enabling potential contributors like @wangyuliu to propagate all the hard work poured into this codebase to new platforms.
